### ISSUE TITLE
ParquetSharp cannot read arrays for large files.

### ISSUE THREAD

**Issue by adeboyed** [G-Research/ParquetSharp#242]
Reading string arrays fails for files which have n rows, where n is > 4096 rows and not a multiple of 4096.
This also occurs for uint arrays, I haven't however tested every Parquet supported type.

### Steps to reproduce
1. Create a Parquet file with a column of string arrays, e.g. using pandas:
```
Python:

import pandas as pd

data = {'data': list(map(lambda x: [f"Row no: {x}"], range(4100))) }
pd.DataFrame(data).to_parquet(r"C:\test.parquet")
```
2. Read Parquet file using ParquetSharp
``` 
C#: 

using var parquetReader = new ParquetFileReader("C:\\test.parquet");
using var rowGroupReader = parquetReader.RowGroup(0);
using var dataColumn = rowGroupReader.Column(0).LogicalReader<string[]>().GetEnumerator();

Console.WriteLine($"No of rows : {parquetReader.FileMetaData.NumRows}");
for (var i = 0; i < rowGroupReader.MetaData.NumRows; i++)
{
   var moveNext = dataColumn.MoveNext();
   Console.WriteLine($"moveNext: {moveNext} data: {dataColumn.Current[0]}");
}
```

**Output:**
```
No of rows : 4100

MoveNext: 'True' | Value: 'Row no: 1'
MoveNext: 'True' | Value: 'Row no: 2'
MoveNext: 'True' | Value: 'Row no: 3'
...
MoveNext: 'True' | Value: 'Row no: 4091'
MoveNext: 'True' | Value: 'Row no: 4092'
MoveNext: 'True' | Value: 'Row no: 4093'
MoveNext: 'True' | Value: 'Row no: 4094'
MoveNext: 'True' | Value: 'Row no: 4095'
MoveNext: 'False' | Value: 'Row no: 4095'
MoveNext: 'False' | Value: 'Row no: 4095'
MoveNext: 'False' | Value: 'Row no: 4095'
MoveNext: 'False' | Value: 'Row no: 4095'
```
As this file has 4100 rows, the value of the single column should keep incrementing until 4100, however after 4096 rows, the enumerator stops returning true for MoveNext(), and ParquetSharp does not return the remaining values.

This is likely related to the default buffer length of 4096, but I haven't properly looked into it it: https://github.com/G-Research/ParquetSharp/blob/39acb40e99bb8e8af5f01d678acefee43d45fbfb/csharp/ColumnReader.cs#L75

Tested versions: 5.0.0-beta1 & 5.0.0-beta2

**Comment by philjdf** [comment]
Thanks for the bug report. Your code calls the following function in ParquetSharp:
```
public IEnumerator<TElement> GetEnumerator()
{
    var buffer = new TElement[BufferLength];

    while (HasNext)
    {
        var read = ReadBatch(buffer);

        for (int i = 0; i != read; ++i)
        {
            yield return buffer[i];
        }
    }
}
```
In the case of arrays, the call to `ReadBatch` does some extra buffering internally. This then causes HasNext to return false (because all of the underlying stream has been read by array reading code), and we don't go round the while loop and refill the buffer.

I guess the immediate user-side fix is to avoid `GetEnumerator`.

Fixing the bug means not using HasNext (so counting how many rows we've read), or making HasNext part of the interface so implementations of `LogicalColumnReader<TElement>` have to implement it like they do for `ReadBatch`.

**Comment by jgiannuzzi** [comment]
Fixed in [ParquetSharp 6.0.1-beta1](https://github.com/G-Research/ParquetSharp/releases/tag/6.0.1-beta1)
